### DISCUSSION TITLE
Extend Devices tab to support device management

### DISCUSSION THREAD

**Discussion by thecode** [home-assistant/architecture#566] [General]
### Background
Major part of Home Assistant is focusing on local control and privacy, allowing users to work with different devices which are not connected to the manufacturer cloud or a 3rd party server. This way of operation has one drawback, most devices will not get firmware updates, even if the device periodically connects to a cloud server to fetch the latest firmware version it will probably need the user to trigger a firmware update which is complicated when the devices are not managed by the manufacture cloud service/app. Most devices with local control support API to check the current firmware version, a status if a new version is available and an option to trigger a firmware update.

### Proposed change
Extend the `Devices` tab and add the following columns:
- Firmware (which will show current firmware version)
- Update available (will show a checkmark if new firmware is available)

Add options to select devices for update or trigger update for all devices.
Device card can be extended to show the new firmware version and also trigger an update for a specific device from the device card.

### Why?
Currently integrations owners & contributors implement sensors to show if a new firmware version is a available, version for the firmware, and services to allow triggering a firmware update, Home Assistant services and sensors should be focused on control and not device management, having an extra sensor to show if a firmware update is available or a service to trigger update clutters the state machine and logbook.  A central solution to manage firmware update will make it a native place to manage updates and will avoid the need for sensors or services. In a later phase we can also discuss if we want to prevent firmware related services/sensors, but the focus of this discussion should be on the implementation of device management from a central place and not the banning of such services or sensors.

#### Pictures of the relevant pages mentioned above that will need to be changed:
![image](https://user-images.githubusercontent.com/1858925/118779954-3b084380-b894-11eb-945e-68b389d739ae.png)
![image](https://user-images.githubusercontent.com/1858925/118780109-63903d80-b894-11eb-85fb-e4983ee23e11.png)

**Comment by MartinHjelmare** [comment]
What are the backend changes required to support the proposal?

**Reply by thecode** [reply]
This are just initial thoughts: 
- Device registry needs to know version of new available firmware and if firmware is upgradable via HA
- API to trigger update with a list of relevant devices

Integrations will need implement new functionality to support this changes

**Comment by janiversen** [comment]
your proposal makes a lot of sense, but it also puts HA in a hot spot, because if HA triggers a firmware update, HA is to some extend responsable.

A natural extension of the API  (in the future) is to actually load the firmware with the all the consequences.

If we go down this road I recommend we make some strong safeguards, ensuring the users are aware of the risk. The reason for my concern is that years ago I helped a great software in the industrial sector and saw it die due to a firmware download mishap.

**Reply by thecode** [reply]
I agree, I was even having thoughts if it should be allowed to trigger FW update from HA, however when official core integrations implement services to update firmware it is more dangerous since services may run in automations without any interaction from the user.
Examples:
https://github.com/home-assistant/core/pull/50390
https://github.com/home-assistant/core/pull/48448

Possible safeguards:
- Do not allow running updates from services/automations only by user confirmation
- Make a popup with user ticking he is aware of the risk (and some legal stuff ...)

**Reply by mib1185** [reply]
I would also limit the functionality to only trigger the update mechanism of the devices it self, but not to upload any firmware files via HA 🤔

**Comment by thecode** [comment]
This PR: https://github.com/home-assistant/core/pull/52932 reminded me about the need for some basic firmware management and information about device connection (in most cases IP address) is something which is frequently requested by users and we end up in integrations making this information as a sensor or an extra state attribute which is totally wrong.

**Comment by codyc1515** [comment]
I think this was actioned with the introduction of the update domain. Can this be closed out now?

**Reply by thecode** [reply]
Actually, although the update domain is the first step to implement the suggestion above, it does not cover my suggestion in the post, but considering there was no response here for 4 years I think this can be closed. I specifically use tools outside HA to do this the update notifications are not useful when a vendor release a new firmware and you have 50 devices popping up with new firmware notifications. The device manager card could be helpful to filter by area for example.
